JavaScript/testing.md

```diff
@@ -0,0 +1,109 @@
+# Testing Philosophy
+
+>>> This was originally written for React Native, but the same principles can be applied to the web.
+
+Bugs come in many shapes and sizes. But in actuality, bugs can be broken down into a few categories.
+
+## Table of Contents
+
+- [Testing Philosophy](#Testing-Philosophy)
+  - [Table of Contents](#Table-of-Contents)
+  - [Bug Types](#Bug-Types)
+  - [Testing Methodologies](#Testing-Methodologies)
+    - [Testing for Style Bugs](#Testing-for-Style-Bugs)
+    - [Testing for Business Requirements](#Testing-for-Business-Requirements)
+    - [Testing for Data Flow](#Testing-for-Data-Flow)
+    - [Testing for Accidental Code Changes](#Testing-for-Accidental-Code-Changes)
+  - [Testing in Summary](#Testing-in-Summary)
+
+<a name="bug-types"></a>
+
+## Bug Types
```

Comment:
one other comes to mind right away, which is when you edit existing code (that is maybe not well tested itself) and your change ends up affecting / breaking something somewhere else that uses it

Reply:
makes sense. do you think that's something that needs to be handled beyond unit tests and making sure that the methods are well typed? can you provide some more color with specific examples? curious how you approach these sorts of problems.

Reply:
I guess this happens when a section of code was not well covered by tests or when we're working on existing codebases with previous authors. If we're following the other best practices then this would be likely less.

Best practice in the case I'm talking about would be to write unit tests before starting to edit the code and research where in the codebase this method / class is used.